added consul and envoy version constraints (#18726)
hashicorp/consul: website/content/docs/connect/config-entries/service-defaults.mdx

    }
    BalanceOutboundConnections = "exact_balance"
    Limits = {
      MaxConnections        = 0
      MaxPendingRequests    = 0
      MaxConcurrentRequests = 0
    }
    PassiveHealthCheck = {
      Interval                = "0s"
      MaxFailures             = 0
      EnforcingConsecutive5xx = 100
      MaxEjectionPercent      = 10
      BaseEjectionTime        = "30s"
    }
  }
  Defaults = {
    Protocol         = "<default protocol for upstream traffic>"
    ConnectTimeoutMs = 5000
    MeshGateway = {
      mode = "<default mode for mesh gateway services>"
    }
    BalanceOutboundConnections = "exact_balance"
    Limits = {
      MaxConnections        = 0
      MaxPendingRequests    = 0
      MaxConcurrentRequests = 0
    }
    PassiveHealthCheck = {
      Interval                = "0s"
      MaxFailures             = 0
      EnforcingConsecutive5xx = 100
      MaxEjectionPercent      = 10
      BaseEjectionTime        = "30s"
    }
  }
}
TransparentProxy = {
  OutboundListenerPort = 15001
  DialedDirectly       = false
}
MutualTLSMode = "strict" # only supported when services are in transparent proxy mode
EnvoyExtensions = [
  {

>     Name = "<name of the extension to implement>"

Reviewer: ```suggestion
    Name = "<name of the extension>"
```

    Required = `false`
    Arguments = { <specific to each extension> }

>     ConsulVersion = "<version of Consul that your extension supports>"
>     EnvoyVersion = "<version of Envoy that your extension supports>"

Reviewer: ```suggestion
    ConsulVersion = "<Consul version constraint for applying the extension>"
    EnvoyVersion = "<Envoy version constraint for applying the extension>"
```

    outboundListenerPort: 15001
    dialedDirectly: false
  mutualTLSMode: strict
  envoyExtensions: 

>   - name: <name of the extension you want to implement>

Reviewer: ```suggestion
  - name: <name of the extension>
```

    required: false
    arguments: 
    - <specific to each extension>

>     consulVersion: <version of Consul that your extension supports>
>     envoyVersion: <version of Envoy that your extension supports> 

Reviewer: ```suggestion
    consulVersion: <Consul version constraint for applying the extension>
    envoyVersion: <Envoy version constraint for applying the extension> 
```

			"Peer": "<peer name of the upstream service>",
			"Protocol": "<protocol for the upstream listener>",
			"ConnectTimeoutMs": 5000,
			"MeshGateway": {
				"Mode": "<mode for mesh gateway services>"
			},
			"BalanceOutboundConnections": "exact_balance",
			"Limits": {
				"MaxConnections": 0,
				"MaxPendingRequests": 0,
				"MaxConcurrentRequests": 0
			},
			"PassiveHealthCheck": {
				"Interval": "0s",
				"MaxFailures": 0,
				"EnforcingConsecutive5xx": 100,
				"MaxEjectionPercent": 10,
				"BaseEjectionTime": "30s"
			}
		}]
	},
	"Defaults": {
		"Protocol": "<protocol for the upstream listener>",
		"ConnectTimeoutMs": 5000,
		"MeshGateway": {
			"Mode": "<mode for mesh gateway services>"
		},
		"BalanceOutboundConnections": "exact_balance",
		"Limits": {
			"MaxConnections": 0,
			"MaxPendingRequests": 0,
			"MaxConcurrentRequests": 0
		},
		"PassiveHealthCheck": {
			"Interval": "0s",
			"MaxFailures": 0,
			"EnforcingConsecutive5xx": 100,
			"MaxEjectionPercent": 10,
			"BaseEjectionTime": "30s"
		}
	},
	"TransparentProxy": {
		"OutboundListenerPort": 15001,
		"DialedDirectly": false
	},
	"MutualTLSMode": "strict",
	"EnvoyExtensions": [{

> 		"Name": "<name of the extension you want to implement",

Reviewer: ```suggestion
		"Name": "<name of the extension>",
```

		"Required": false,
		"Arguments": {
			"<specific to each extension>": "<specific to each extension>"
		},

> 		"ConsulVersion": "<version of Consul that your extension supports>",
> 		"EnvoyVersion": "<version of Envoy that your extension supports>"

Reviewer: ```suggestion
		"ConsulVersion": "<Consul version constraint for applying the extension>",
		"EnvoyVersion": "<Envoy version constraint for applying the extension>"
```

List of extensions to modify Envoy proxy configuration. Refer to [Envoy Extensions](/consul/docs/connect/proxies/envoy-extensions) for additional information.

The following table describes how to configure values in the `EnvoyExtensions` map:

| Parameter | Description | Data type | Default |
| ---       | ---         | ---       | ---     |
| `Name` | Specifies the name of the extension. | String | None |

> | `Required` | Specify `true` to require Consul to implement the extension. <p>Use this parameter to ensure that extensions required for secure communication are not unintentionally bypassed.</p><p>Consul returns an error when `Required` is set to `true` and the extension does not update Envoy resources. </p> | String | None |

Reviewer: Consul doesn't so much _implement_ an extension as apply code that's already there and check whether application is successful; tweaked language to match. Also clarified the effect of errors at runtime.

```suggestion
| `Required` | Specify `true` to require the extension to apply successfully. <p>Use this parameter to ensure that extensions required for secure communication are not unintentionally bypassed.</p><p>Consul logs an error and does not update xDS resources (skipping _all_ extensions) when an extension with `Required` set to `true` fails to apply. </p> | String | None |
```

> | `Arguments` | Specifies the arguments to pass to the extension executable. Refer to the documentation for the extension you want to implement for additional information. | Map | None |

Reviewer: We don't currently support external extensions, so I think this might have been unintentionally included from a future-facing part of an earlier RFC.
```suggestion
| `Arguments` | Specifies the arguments to pass to the extension. Refer to the documentation for the extension you want to implement for additional information. | Map | None |
```

> | `ConsulVersion` | Specifies the version of Consul that the extension is allowed to work with. Consul validates the version during xDS updates. If a different version is in use, Consul skips the extension and writes the event to the log. <p>The `ConsulVersion` and `EnvoyVersion` must both validate for Consul to implement the extension.</p> | String | None |
> | `EnvoyVersion` | Specifies the version of Envoy that the extension is allowed to work with. Consul validates the version during xDS updates. If a different version is in use, Consul skips the extension and writes the event to the log. <p>The `ConsulVersion` and `EnvoyVersion` must both validate for Consul to implement the extension.</p> | String | None |

Reviewer: A few suggestions, feel free to edit:
* Linked to `go-version` library for documentation on how to use the version constraint syntax (it can be an interpreted range for flexibility, not just a literal match)
* Clarified how the versions are determined
* Added note on when to use

IMO the last bit about both passing isn't super necessary since it's implied by the failure behavior note (this is just another, optional way for validation/application to "fail").

```suggestion
| `ConsulVersion` | Specifies the Consul [version constraint](https://github.com/hashicorp/go-version) for the extension. Consul validates this against its runtime version during xDS updates. If a non-matching version is in use, Consul will log and skip the extension. <p>Use this parameter to avoid upgrade issues when a configured extension is not compatible with a new version of Consul.</p> | String | None |
| `EnvoyVersion` | Specifies the Envoy [version constraint](https://github.com/hashicorp/go-version) for the extension. Consul validates this against the version of the running Envoy proxy during xDS updates. If a non-matching version is in use, Consul will log and skip the extension. <p>Use this parameter to avoid upgrade issues when a configured extension is not compatible with a new version of Envoy.</p> | String | None |
```

#### Values

The following table describes how to configure values in the `envoyExtensions` map:

| Parameter | Description | Data type | Default |
| ---       | ---         | ---       | ---     |
| `name` | Specifies the name of the extension. | String | None |

> | `required` | Specify `true` to require Consul to implement the extension. <p>Use this parameter to ensure that extensions required for secure communication are not unintentionally bypassed.</p><p>Consul returns an error when `required` is set to `true` and the extension does not update Envoy resources. </p> | String | None |

Reviewer: ```suggestion
| `required` | Specify `true` to require the extension to apply successfully. <p>Use this parameter to ensure that extensions required for secure communication are not unintentionally bypassed.</p><p>Consul logs an error and does not update xDS resources (skipping _all_ extensions) when an extension with `Required` set to `true` fails to apply. </p> | String | None |
```

> | `arguments` | Specifies the arguments to pass to the extension executable. Refer to the documentation for the extension you want to implement for additional information. | Map | None |

Reviewer: ```suggestion
| `arguments` | Specifies the arguments to pass to the extension. Refer to the documentation for the extension for additional information. | Map | None |
```

> | `consulVersion` | Specifies the version of Consul that the extension is allowed to work with. Consul validates the version during xDS updates. If a different version is in use, Consul skips the extension and writes the event to the log. <p>The `consulVersion` and `envoyVersion` must both validate for Consul to implement the extension.</p> | String | None |
> | `envoyVersion` | Specifies the version of Envoy that the extension is allowed to work with. Consul validates the version during xDS updates. If a different version is in use, Consul skips the extension and writes the event to the log. <p>The `consulVersion` and `envoyVersion` must both validate for Consul to implement the extension.</p> | String | None |

Reviewer: ```suggestion
| `consulVersion` | Specifies the Consul [version constraint](https://github.com/hashicorp/go-version) for the extension. Consul validates this against its runtime version during xDS updates. If a non-matching version is in use, Consul will log and skip the extension. <p>Use this parameter to avoid upgrade issues when a configured extension is not compatible with a new version of Consul.</p> | String | None |
| `envoyVersion` | Specifies the Envoy [version constraint](https://github.com/hashicorp/go-version) for the extension. Consul validates this against the version of the running Envoy proxy during xDS updates. If a non-matching version is in use, Consul will log and skip the extension. <p>Use this parameter to avoid upgrade issues when a configured extension is not compatible with a new version of Envoy.</p> | String | None |
```
